File: src/Services/VarianceVibrationService.php
<?php
/**
 * Variance Vibration Service
 * PHP 7.1.9 Compatible Service for Variance Vibration Feature
 * Integrates with Moodle 3.7 and MySQL 5.7
 */

namespace Alt42\Services;

use Alt42\Database\Connection;
use PDO;
use PDOException;
use Exception;

class VarianceVibrationService
{
    /**
     * @var PDO Database connection
     */
    private $db;

    /**
     * @var CacheService Cache service instance
     */
    private $cache;

    /**
     * Constructor
     *
     * @param PDO|null $db Database connection
     * @param CacheService|null $cache Cache service
     */
    public function __construct($db = null, $cache = null)
    {
        $this->db = $db ?: Connection::getInstance()->getConnection();
        $this->cache = $cache;
    }

    /**
     * Get variance problem by ID
     *
     * @param string $problemId Problem ID
     * @return array|null Problem data or null if not found
     */
    public function getProblemById($problemId)
    {
        try {
            $sql = "SELECT * FROM variance_problems WHERE id = :id AND is_active = 1";
            $stmt = $this->db->prepare($sql);
            $stmt->execute(['id' => $problemId]);

            $problem = $stmt->fetch(PDO::FETCH_ASSOC);

            if (!$problem) {
                return null;
            }

            return $this->formatProblem($problem);
        } catch (PDOException $e) {
            error_log("Error fetching variance problem: " . $e->getMessage());
            return null;
        }
    }

    /**
     * Get problems by category and difficulty
     *
     * @param string $category Problem category
     * @param string|null $difficulty Difficulty level (easy, medium, hard)
     * @param int $limit Maximum number of problems
     * @param int $offset Offset for pagination
     * @return array Array of problems
     */
    public function getProblemsByCategory($category, $difficulty = null, $limit = 10, $offset = 0)
    {
        try {
            $sql = "SELECT * FROM variance_problems WHERE category = :category AND is_active = 1";
            $params = ['category' => $category];

            if ($difficulty !== null) {
                $sql .= " AND difficulty = :difficulty";
                $params['difficulty'] = $difficulty;
            }

            $sql .= " ORDER BY created_at DESC LIMIT :limit OFFSET :offset";

            $stmt = $this->db->prepare($sql);

            // Bind parameters
            foreach ($params as $key => $value) {
                $stmt->bindValue(':' . $key, $value);
            }
            $stmt->bindValue(':limit', (int)$limit, PDO::PARAM_INT);
            $stmt->bindValue(':offset', (int)$offset, PDO::PARAM_INT);

            $stmt->execute();
            $problems = $stmt->fetchAll(PDO::FETCH_ASSOC);

            return array_map([$this, 'formatProblem'], $problems);
        } catch (PDOException $e) {
            error_log("Error fetching variance problems: " . $e->getMessage());
            return [];
        }
    }

    /**
     * Calculate variance statistics for a dataset
     *
     * @param array $values Array of numeric values
     * @param bool $isSample Use sample variance (n-1) or population (n)
     * @return array Variance statistics
     */
    public function calculateVarianceStats($values, $isSample = false)
    {
        if (empty($values)) {
            return [
                'mean' => 0,
                'variance' => 0,
                'standardDeviation' => 0,
                'count' => 0,
                'min' => 0,
                'max' => 0,
                'range' => 0,
            ];
        }

        $count = count($values);
        $mean = array_sum($values) / $count;

        // Calculate squared differences
        $squaredDiffs = array_map(function($value) use ($mean) {
            return pow($value - $mean, 2);
        }, $values);

        $sumSquaredDiffs = array_sum($squaredDiffs);

        // Calculate variance
        $divisor = $isSample ? ($count - 1) : $count;
        $variance = $divisor > 0 ? $sumSquaredDiffs / $divisor : 0;

        $standardDeviation = sqrt($variance);
        $min = min($values);
        $max = max($values);

        return [
            'mean' => round($mean, 6),
            'variance' => round($variance, 6),
            'standardDeviation' => round($standardDeviation, 6),
            'count' => $count,
            'min' => $min,
            'max' => $max,
            'range' => $max - $min,
        ];
    }

    /**
     * Map variance to vibration intensity (1-10)
     *
     * @param float $variance Calculated variance
     * @param float $mean Mean of dataset
     * @param array $config Vibration configuration
     * @return int Intensity level (1-10)
     */
    public function varianceToIntensity($variance, $mean, $config = [])
    {
        $defaults = [
            'minIntensity' => 1,
            'maxIntensity' => 10,
            'varianceThreshold' => 100.0,
            'useNormalizedVariance' => true,
        ];

        $config = array_merge($defaults, $config);

        if ($config['useNormalizedVariance'] && $mean != 0) {
            // Coefficient of variation approach
            $cv = sqrt($variance) / abs($mean);
            $normalizedValue = min($cv / 2.0, 1.0);
        } else {
            // Raw variance with threshold
            $normalizedValue = min($variance / $config['varianceThreshold'], 1.0);
        }

        // Map to intensity range
        $intensityRange = $config['maxIntensity'] - $config['minIntensity'];
        $intensity = $config['minIntensity'] + ($normalizedValue * $intensityRange);

        return max($config['minIntensity'], min($config['maxIntensity'], round($intensity)));
    }

    /**
     * Generate vibration pattern based on intensity
     *
     * @param int $intensity Vibration intensity (1-10)
     * @param string $patternType Type of pattern (continuous, pulsed, progressive)
     * @return array Array of vibration pulse durations in ms
     */
    public function generateVibrationPattern($intensity, $patternType = 'progressive')
    {
        $baseDuration = 50 + ($intensity * 15); // 50-200ms

        switch ($patternType) {
            case 'continuous':
                return [$baseDuration * 2];

            case 'pulsed':
                $pulseCount = ceil($intensity / 3); // 1-4 pulses
                $pattern = [];
                for ($i = 0; $i < $pulseCount; $i++) {
                    $pattern[] = $baseDuration;
                    if ($i < $pulseCount - 1) {
                        $pattern[] = 50; // Gap
                    }
                }
                return $pattern;

            case 'progressive':
                $steps = min($intensity, 5);
                $pattern = [];
                for ($i = 1; $i <= $steps; $i++) {
                    $pattern[] = (int)($baseDuration * ($i / $steps));
                    if ($i < $steps) {
                        $pattern[] = 30; // Short gap
                    }
                }
                return $pattern;

            default:
                return [$baseDuration];
        }
    }

    /**
     * Submit student answer
     *
     * @param string $problemId Problem ID
     * @param string $studentId Student ID
     * @param float $submittedAnswer Student's answer
     * @param int $timeSpentSeconds Time spent on problem
     * @param array $deviceInfo Device information
     * @return array Submission result
     */
    public function submitAnswer($problemId, $studentId, $submittedAnswer, $timeSpentSeconds, $deviceInfo = [])
    {
        try {
            // Get problem
            $problem = $this->getProblemById($problemId);
            if (!$problem) {
                throw new Exception('Problem not found');
            }

            // Validate answer
            $tolerance = $this->getToleranceForDifficulty($problem['difficulty']);
            $expectedVariance = $problem['expectedVariance'];

            $percentError = $expectedVariance != 0
                ? abs($submittedAnswer - $expectedVariance) / $expectedVariance * 100
                : ($submittedAnswer == 0 ? 0 : 100);

            $isCorrect = $percentError <= ($tolerance * 100);

            // Calculate vibration feedback
            $config = $this->getVibrationConfig();
            $stats = $this->calculateVarianceStats($problem['dataSet'], $problem['isSample']);
            $intensity = $this->varianceToIntensity($stats['variance'], $stats['mean'], $config);
            $pattern = $this->generateVibrationPattern($intensity, $config['patternType']);

            // Get attempt number
            $attemptNumber = $this->getAttemptNumber($studentId, $problemId) + 1;

            // Create submission record
            $submissionId = $this->generateUuid();

            $sql = "INSERT INTO variance_submissions (
                id, problem_id, student_id, submitted_answer, is_correct,
                actual_variance, percent_error, vibration_intensity,
                vibration_pattern, vibration_triggered, time_spent_seconds,
                attempt_number, device_info, submitted_at
            ) VALUES (
                :id, :problem_id, :student_id, :submitted_answer, :is_correct,
                :actual_variance, :percent_error, :vibration_intensity,
                :vibration_pattern, :vibration_triggered, :time_spent_seconds,
                :attempt_number, :device_info, NOW()
            )";

            $stmt = $this->db->prepare($sql);
            $stmt->execute([
                'id' => $submissionId,
                'problem_id' => $problemId,
                'student_id' => $studentId,
                'submitted_answer' => $submittedAnswer,
                'is_correct' => $isCorrect ? 1 : 0,
                'actual_variance' => $expectedVariance,
                'percent_error' => $percentError,
                'vibration_intensity' => $intensity,
                'vibration_pattern' => json_encode($pattern),
                'vibration_triggered' => 1,
                'time_spent_seconds' => $timeSpentSeconds,
                'attempt_number' => $attemptNumber,
                'device_info' => json_encode($deviceInfo),
            ]);

            return [
                'submissionId' => $submissionId,
                'isCorrect' => $isCorrect,
                'percentError' => round($percentError, 2),
                'expectedVariance' => $expectedVariance,
                'vibrationIntensity' => $intensity,
                'vibrationPattern' => $pattern,
                'attemptNumber' => $attemptNumber,
            ];

        } catch (Exception $e) {
            error_log("Error submitting answer: " . $e->getMessage());
            throw $e;
        }
    }

    /**
     * Get student's attempt number for a problem
     *
     * @param string $studentId Student ID
     * @param string $problemId Problem ID
     * @return int Attempt count
     */
    private function getAttemptNumber($studentId, $problemId)
    {
        $sql = "SELECT COUNT(*) as count FROM variance_submissions
                WHERE student_id = :student_id AND problem_id = :problem_id";

        $stmt = $this->db->prepare($sql);
        $stmt->execute([
            'student_id' => $studentId,
            'problem_id' => $problemId,
        ]);

        $result = $stmt->fetch(PDO::FETCH_ASSOC);
        return (int)$result['count'];
    }

    /**
     * Get tolerance based on difficulty level
     *
     * @param string $difficulty Difficulty level
     * @return float Tolerance percentage (0-1)
     */
    private function getToleranceForDifficulty($difficulty)
    {
        $tolerances = [
            'easy' => 0.15,
            'medium' => 0.10,
            'hard' => 0.05,
        ];

        return isset($tolerances[$difficulty]) ? $tolerances[$difficulty] : 0.10;
    }

    /**
     * Get vibration configuration
     *
     * @return array Configuration array
     */
    private function getVibrationConfig()
    {
        try {
            $sql = "SELECT setting_value FROM variance_vibration_settings
                    WHERE setting_key = 'default_vibration_config' AND is_active = 1";

            $stmt = $this->db->prepare($sql);
            $stmt->execute();

            $result = $stmt->fetch(PDO::FETCH_ASSOC);

            if ($result) {
                return json_decode($result['setting_value'], true);
            }
        } catch (PDOException $e) {
            error_log("Error fetching vibration config: " . $e->getMessage());
        }

        // Return defaults if not found
        return [
            'minIntensity' => 1,
            'maxIntensity' => 10,
            'varianceThreshold' => 100.0,
            'useNormalizedVariance' => true,
            'vibrationEnabled' => true,
            'patternType' => 'progressive',
            'debounceMs' => 300,
        ];
    }

    /**
     * Format problem data for API response
     *
     * @param array $problem Raw problem data from database
     * @return array Formatted problem data
     */
    private function formatProblem($problem)
    {
        return [
            'id' => $problem['id'],
            'questionText' => $problem['question_text'],
            'questionTextKo' => $problem['question_text_ko'],
            'dataSet' => json_decode($problem['data_set'], true),
            'isSample' => (bool)$problem['is_sample'],
            'expectedVariance' => (float)$problem['expected_variance'],
            'expectedStdDev' => (float)$problem['expected_std_dev'],
            'meanValue' => (float)$problem['mean_value'],
            'hints' => json_decode($problem['hints'], true),
            'hintsKo' => json_decode($problem['hints_ko'], true),
            'category' => $problem['category'],
            'difficulty' => $problem['difficulty'],
            'gradeLevel' => $problem['grade_level'],
            'tags' => json_decode($problem['tags'], true),
        ];
    }

    /**
     * Generate UUID v4
     *
     * @return string UUID string
     */
    private function generateUuid()
    {
        return sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff)
        );
    }
}

File: backend/services/variance_vibration_service.py
"""
Variance Vibration Service
Backend service for managing variance problems and vibration feedback
Integrates with Moodle LMS and provides API for frontend components
"""

from typing import List, Dict, Optional, Any
from dataclasses import dataclass, asdict
from datetime import datetime
from enum import Enum
import statistics
import math
import json
import asyncio
from fastapi import HTTPException


class VarianceType(str, Enum):
    """Type of variance calculation"""
    POPULATION = "population"
    SAMPLE = "sample"


class DifficultyLevel(str, Enum):
    """Difficulty level for variance problems"""
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


@dataclass
class VarianceStats:
    """Statistical calculations for a dataset"""
    mean: float
    variance: float
    standard_deviation: float
    count: int
    min_value: float
    max_value: float
    range_value: float
    variance_type: VarianceType


@dataclass
class VarianceProblem:
    """Variance problem definition"""
    id: str
    question_text: str
    question_text_ko: str
    data_set: List[float]
    is_sample: bool
    expected_variance: float
    hints: List[str]
    hints_ko: List[str]
    category: str
    difficulty: DifficultyLevel
    moodle_question_id: Optional[int] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


@dataclass
class VibrationConfig:
    """Vibration configuration for variance feedback"""
    min_intensity: int = 1
    max_intensity: int = 10
    variance_threshold: float = 100.0
    use_normalized_variance: bool = True
    vibration_enabled: bool = True
    pattern_type: str = "progressive"  # continuous, pulsed, progressive
    debounce_ms: int = 300


@dataclass
class VarianceSubmission:
    """Student's variance problem submission"""
    id: str
    problem_id: str
    student_id: str
    submitted_answer: float
    is_correct: bool
    actual_variance: float
    percent_error: float
    vibration_intensity: int
    vibration_pattern: List[int]
    time_spent_seconds: int
    submitted_at: datetime


class VarianceCalculator:
    """Utility class for variance calculations"""

    @staticmethod
    def calculate_mean(values: List[float]) -> float:
        """Calculate the mean of values"""
        if not values:
            return 0.0
        return statistics.mean(values)

    @staticmethod
    def calculate_variance(
        values: List[float],
        variance_type: VarianceType = VarianceType.POPULATION
    ) -> float:
        """Calculate variance (population or sample)"""
        if not values:
            return 0.0
        if len(values) == 1 and variance_type == VarianceType.SAMPLE:
            return 0.0

        if variance_type == VarianceType.SAMPLE:
            return statistics.variance(values)
        else:
            return statistics.pvariance(values)

    @staticmethod
    def calculate_standard_deviation(
        values: List[float],
        variance_type: VarianceType = VarianceType.POPULATION
    ) -> float:
        """Calculate standard deviation"""
        if not values:
            return 0.0

        if variance_type == VarianceType.SAMPLE:
            return statistics.stdev(values)
        else:
            return statistics.pstdev(values)

    @staticmethod
    def calculate_stats(
        values: List[float],
        variance_type: VarianceType = VarianceType.POPULATION
    ) -> VarianceStats:
        """Calculate comprehensive variance statistics"""
        if not values:
            return VarianceStats(
                mean=0.0,
                variance=0.0,
                standard_deviation=0.0,
                count=0,
                min_value=0.0,
                max_value=0.0,
                range_value=0.0,
                variance_type=variance_type
            )

        mean = VarianceCalculator.calculate_mean(values)
        variance = VarianceCalculator.calculate_variance(values, variance_type)
        std_dev = math.sqrt(variance)
        min_val = min(values)
        max_val = max(values)

        return VarianceStats(
            mean=mean,
            variance=variance,
            standard_deviation=std_dev,
            count=len(values),
            min_value=min_val,
            max_value=max_val,
            range_value=max_val - min_val,
            variance_type=variance_type
        )

    @staticmethod
    def normalize_variance(variance: float, mean: float) -> float:
        """
        Normalize variance to 0-1 scale using coefficient of variation
        CV = σ/μ, normalized to 0-1 range (capped at CV=2.0)
        """
        if mean == 0:
            return 0.0

        cv = math.sqrt(variance) / abs(mean)
        normalized_cv = min(cv / 2.0, 1.0)

        return normalized_cv

    @staticmethod
    def validate_answer(
        submitted: float,
        expected: float,
        tolerance: float = 0.1
    ) -> tuple[bool, float]:
        """
        Validate student's answer
        Returns (is_correct, percent_error)
        """
        if expected == 0:
            return (submitted == 0, 0.0 if submitted == 0 else 100.0)

        percent_error = abs(submitted - expected) / expected
        is_correct = percent_error <= tolerance

        return (is_correct, percent_error * 100)


class VibrationMapper:
    """Maps variance values to vibration intensity and patterns"""

    @staticmethod
    def variance_to_intensity(
        variance: float,
        mean: float,
        config: VibrationConfig
    ) -> int:
        """Map variance to vibration intensity (1-10)"""
        if config.use_normalized_variance and mean != 0:
            normalized_value = VarianceCalculator.normalize_variance(variance, mean)
        else:
            normalized_value = min(variance / config.variance_threshold, 1.0)

        # Map to intensity range
        intensity_range = config.max_intensity - config.min_intensity
        intensity = config.min_intensity + normalized_value * intensity_range

        # Round and clamp
        return max(config.min_intensity, min(config.max_intensity, round(intensity)))

    @staticmethod
    def generate_vibration_pattern(
        intensity: int,
        pattern_type: str = "progressive"
    ) -> List[int]:
        """Generate vibration pattern based on intensity"""
        base_duration = 50 + intensity * 15  # 50-200ms

        if pattern_type == "continuous":
            return [base_duration * 2]

        elif pattern_type == "pulsed":
            pulse_count = math.ceil(intensity / 3)  # 1-4 pulses
            pattern = []
            for i in range(pulse_count):
                pattern.append(base_duration)
                if i < pulse_count - 1:
                    pattern.append(50)  # Gap
            return pattern

        elif pattern_type == "progressive":
            steps = min(intensity, 5)
            pattern = []
            for i in range(1, steps + 1):
                pattern.append(int(base_duration * (i / steps)))
                if i < steps:
                    pattern.append(30)  # Short gap
            return pattern

        else:
            return [base_duration]

    @staticmethod
    def get_intensity_description(intensity: int, language: str = "en") -> str:
        """Get human-readable intensity description"""
        descriptions = {
            "en": {
                1: "Very Low", 2: "Very Low",
                3: "Low", 4: "Low",
                5: "Medium", 6: "Medium",
                7: "High", 8: "High",
                9: "Very High", 10: "Very High"
            },
            "ko": {
                1: "매우 낮음", 2: "매우 낮음",
                3: "낮음", 4: "낮음",
                5: "보통", 6: "보통",
                7: "높음", 8: "높음",
                9: "매우 높음", 10: "매우 높음"
            }
        }

        return descriptions.get(language, descriptions["en"]).get(intensity, "Unknown")


class VarianceVibrationService:
    """Main service for variance vibration feature"""

    def __init__(self, db_connection=None):
        self.db = db_connection
        self.calculator = VarianceCalculator()
        self.mapper = VibrationMapper()

    async def get_problem_by_id(self, problem_id: str) -> Optional[VarianceProblem]:
        """Retrieve variance problem by ID"""
        if not self.db:
            # Return sample problem for testing
            return self._get_sample_problem(problem_id)

        # TODO: Implement database query
        query = """
            SELECT id, question_text, question_text_ko, data_set, is_sample,
                   expected_variance, hints, hints_ko, category, difficulty,
                   moodle_question_id, created_at, updated_at
            FROM variance_problems
            WHERE id = $1
        """
        # result = await self.db.fetchrow(query, problem_id)
        # return self._parse_problem(result)

        return self._get_sample_problem(problem_id)

    async def get_problems_by_category(
        self,
        category: str,
        difficulty: Optional[DifficultyLevel] = None,
        limit: int = 10
    ) -> List[VarianceProblem]:
        """Retrieve problems by category and difficulty"""
        # TODO: Implement database query
        return [self._get_sample_problem(f"{category}_{i}") for i in range(min(3, limit))]

    async def calculate_variance_stats(
        self,
        values: List[float],
        is_sample: bool = False
    ) -> VarianceStats:
        """Calculate variance statistics for a dataset"""
        variance_type = VarianceType.SAMPLE if is_sample else VarianceType.POPULATION
        return self.calculator.calculate_stats(values, variance_type)

    async def generate_vibration_config(
        self,
        values: List[float],
        is_sample: bool = False,
        config: Optional[VibrationConfig] = None
    ) -> Dict[str, Any]:
        """Generate vibration configuration for given data"""
        if config is None:
            config = VibrationConfig()

        variance_type = VarianceType.SAMPLE if is_sample else VarianceType.POPULATION
        stats = self.calculator.calculate_stats(values, variance_type)

        intensity = self.mapper.variance_to_intensity(
            stats.variance,
            stats.mean,
            config
        )

        pattern = self.mapper.generate_vibration_pattern(
            intensity,
            config.pattern_type
        )

        return {
            "stats": asdict(stats),
            "intensity": intensity,
            "vibration_pattern": pattern,
            "normalized_variance": VarianceCalculator.normalize_variance(
                stats.variance,
                stats.mean
            ),
            "intensity_description_en": self.mapper.get_intensity_description(
                intensity, "en"
            ),
            "intensity_description_ko": self.mapper.get_intensity_description(
                intensity, "ko"
            ),
            "config": asdict(config)
        }

    async def submit_answer(
        self,
        problem_id: str,
        student_id: str,
        submitted_answer: float,
        time_spent_seconds: int
    ) -> VarianceSubmission:
        """Process student's answer submission"""
        problem = await self.get_problem_by_id(problem_id)
        if not problem:
            raise HTTPException(status_code=404, detail="Problem not found")

        is_correct, percent_error = self.calculator.validate_answer(
            submitted_answer,
            problem.expected_variance
        )

        # Generate vibration feedback
        vibration_config = VibrationConfig()
        stats = await self.calculate_variance_stats(
            problem.data_set,
            problem.is_sample
        )

        intensity = self.mapper.variance_to_intensity(
            stats.variance,
            stats.mean,
            vibration_config
        )

        pattern = self.mapper.generate_vibration_pattern(
            intensity,
            vibration_config.pattern_type
        )

        submission = VarianceSubmission(
            id=f"sub_{datetime.utcnow().timestamp()}",
            problem_id=problem_id,
            student_id=student_id,
            submitted_answer=submitted_answer,
            is_correct=is_correct,
            actual_variance=problem.expected_variance,
            percent_error=percent_error,
            vibration_intensity=intensity,
            vibration_pattern=pattern,
            time_spent_seconds=time_spent_seconds,
            submitted_at=datetime.utcnow()
        )

        # TODO: Save to database
        # await self._save_submission(submission)

        return submission

    def _get_sample_problem(self, problem_id: str) -> VarianceProblem:
        """Generate sample problem for testing"""
        sample_data = [10.0, 12.0, 11.0, 13.0, 12.0, 10.0, 14.0, 11.0]
        variance = self.calculator.calculate_variance(sample_data, VarianceType.SAMPLE)

        return VarianceProblem(
            id=problem_id,
            question_text="Calculate the variance of the following dataset:",
            question_text_ko="다음 데이터의 분산을 계산하세요:",
            data_set=sample_data,
            is_sample=True,
            expected_variance=variance,
            hints=[
                "First, calculate the mean of all values",
                "Then, find the squared difference from mean for each value",
                "Finally, divide by (n-1) for sample variance"
            ],
            hints_ko=[
                "먼저 모든 값의 평균을 계산하세요",
                "그런 다음 각 값에서 평균을 뺀 값의 제곱을 구하세요",
                "마지막으로 (n-1)로 나누어 표본 분산을 구하세요"
            ],
            category="basic_statistics",
            difficulty=DifficultyLevel.EASY,
            moodle_question_id=None,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )


# Export for API usage
__all__ = [
    'VarianceVibrationService',
    'VarianceProblem',
    'VarianceStats',
    'VibrationConfig',
    'VarianceSubmission',
    'VarianceType',
    'DifficultyLevel',
]
